allennlp/training/trainer.py

```diff
@@ -282,7 +290,14 @@ def __init__(
             self._pytorch_model = self.model
 
     def rescale_gradients(self) -> Optional[float]:
-        return training_util.rescale_gradients(self.model, self._grad_norm)
+        if self._opt_level is not None:
+            if self._grad_norm:
+                # See: https://nvidia.github.io/apex/advanced.html#gradient-clipping
+                parameters_to_clip = [p for p in amp.master_params(self.optimizer) if p.grad is not None]
+                return training_util.sparse_clip_norm(parameters_to_clip, self._grad_norm)
+            return None
+        else:
+            return training_util.rescale_gradients(self.model, self._grad_norm)
```

Comment:
Another option here is to add a `use_amp` flag to `rescale_gradients`.  Oh, except that then also requires passing in the optimizer...  Yeah, I think what you have here is good enough.

Reply:
Yes exactly my thinking. Othewise, we could have a second util function named something like `rescale_gradients_amp` which accepts an optimizer.

Reply:
I think adding that function would be overkill.  If we want to refactor things, we should have `rescale_gradients` take a list of parameters.  But then we can just remove it and call `sparse_clip_norm` directly.  I'm somewhat in favor of that option, as it would simplify the above code.

Reply:
Okay, I refactored `Trainer.rescale_gradients` to use `training_utils.sparse_clip_norm` directly. Otherwise the function is the same.